src/utils/generator.js

```diff
@@ -94,15 +243,65 @@ export class Generator {
     this._entries = [];
     this.id = realm.nextGeneratorId++;
     this._name = name;
+    this.effectsToApply = effects;
   }
 
   realm: Realm;
   _entries: Array<GeneratorEntry>;
   preludeGenerator: PreludeGenerator;
+  effectsToApply: void | Effects;
 
   id: number;
   _name: string;
 
+  // Make sure to to fixup
+  // how to apply things around sets of things
+  static fromEffects(effects: Effects, realm: Realm): Generator {
+    let [result, generator, modifiedBindings, modifiedProperties, createdObjects] = effects;
+
+    let output = new Generator(realm, "AdditionalFunctionEffects" + additionalFunctionUID++, effects);
+    output.appendGenerator(generator, "Additional function generator");
+
+    for (let [propertyBinding, newValue] of modifiedProperties) {
+      let object = propertyBinding.object;
+      if (object instanceof ObjectValue && createdObjects.has(object)) continue; // Created Object's binding
+      if (object.refuseSerialization) continue; // modification to internal state
+      if (object.isIntrinsic()) continue; // Avoid double-counting
+      output.appendPropertyModification(propertyBinding, newValue);
+    }
+
+    for (let [modifiedBinding, newValue] of modifiedBindings) {
+      output.appendBindingModification(modifiedBinding, newValue);
+    }
+
+    if (!(result instanceof UndefinedValue)) output.appendReturnValue(result);
+    return output;
+  }
+
+  appendPropertyModification(propertyBinding: PropertyBinding, newValue: void | Descriptor) {
```

Comment:
The other function that push single a entry all have the naming scheme `emit...`. Let's keep it consistent.